### PR TITLE
adapter: Generate privileges for SQL statement

### PR DESCRIPTION
**Pull request by jkosh44**
This commit adds logic to generate the required privileges for
executing some SQL statement. It also updates the RBAC design doc to
match this logic.
    
This commit doesn't actually check the generated privileges before,
executing a statement. That is left for a later commit.
    
Part of MaterializeInc/database-issues#3380

### Motivation
This PR adds a known-desirable feature.

### Checklist

- [X] This PR has adequate test coverage / QA involvement has been duly considered.
- [X] This PR has an associated up-to-date [design doc](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/design/README.md), is a design doc ([template](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/design/00000000_template.md)), or is sufficiently small to not require a design.
  
- [ ] This PR evolves [an existing `$T ⇔ Proto$T` mapping](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/command-and-response-binary-encoding.md) (possibly in a backwards-incompatible way) and therefore is tagged with a `T-proto` label.
- [ ] If this PR will require changes to cloud orchestration, there is a companion cloud PR to account for those changes that is tagged with the release-blocker label ([example](https://github.com/MaterializeInc/cloud/pull/5021)).
  
- [ ] This PR includes the following [user-facing behavior changes](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/guide-changes.md#what-changes-require-a-release-note):
  -

### CODE REVIEW

**Comment by jkosh44** [comment]
The code added in this PR is executed but the result is thrown out and never used. The code is only executed if RBAC is enabled, so we don't have to worry about it being run on actual customer environments. The code is run as part of the existing RBAC tests so we know there's no major panics, but there are no tests that check the result of the new code since it's thrown out.

I thought it may still be useful to include this in a separate PR so the docs update and the privilege generation could be looked at in isolation, even if there may be some subtle untested correctness issues. The next PR will actually look at the generated privileges and add a lot of tests. If people disagree then I'm happy to close this and fold it into the next PR.

File: src/sql/src/plan.rs [hunk 1]
```diff
@@ -105,7 +105,7 @@ pub enum Plan {
     DropObjects(DropObjectsPlan),
     EmptyQuery,
     ShowAllVariables,
-    ShowCreate(SendRowsPlan),
```
Comment (mjibson) [inline src/sql/src/plan.rs:108]: Why did these changes need to be in this PR?

File: src/adapter/src/rbac.rs [hunk 2]
```diff
@@ -204,15 +214,35 @@ fn generate_required_plan_attribute(plan: &Plan) -> Option<Attribute> {
...
             Some(Attribute::CreateRole)
         }
-        Plan::CreateSource(_)
-        | Plan::CreateSources(_)
-        | Plan::CreateTable(_)
+        Plan::CreateSource(CreateSourcePlan { cluster_config, .. })
+        | Plan::CreateSink(CreateSinkPlan { cluster_config, .. }) => {
+            if matches!(
+                cluster_config,
+                SourceSinkClusterConfig::Linked { .. } | SourceSinkClusterConfig::Undefined
```
Comment (mjibson) [inline src/adapter/src/rbac.rs:221]: I don't understand why the `Existing` variant doesn't require the create cluster attribute. (This is my first time seeing SourceSinkClusterConfig though, so I'm probably missing something obvious.)

File: src/adapter/src/rbac.rs [hunk 3]
```diff
@@ -204,15 +214,35 @@ fn generate_required_plan_attribute(plan: &Plan) -> Option<Attribute> {
         Plan::DropObjects(plan) if plan.object_type == ObjectType::Role => {
             Some(Attribute::CreateRole)
         }
-        Plan::CreateSource(_)
-        | Plan::CreateSources(_)
-        | Plan::CreateTable(_)
+        Plan::CreateSource(CreateSourcePlan { cluster_config, .. })
```
Comment (mjibson) [inline src/adapter/src/rbac.rs:217]: Is this diff related to required privs? It looks like it's related to the test below, and they could be in another commit?

File: src/adapter/src/rbac.rs [hunk 4]
```diff
@@ -418,6 +448,417 @@ fn ownership_err(
...
+    role_id: RoleId,
+) -> BTreeSet<(ObjectId, AclMode, RoleId)> {
+    match plan {
+        Plan::CreateConnection(plan) => {
+            let mut privileges = BTreeSet::from([(
+                plan.name.qualifiers.clone().into(),
+                AclMode::CREATE,
+                role_id,
+            )]);
+            privileges.extend(generate_item_usage_privileges(catalog, depends_on, role_id));
```
Comment (mjibson) [inline src/adapter/src/rbac.rs:469]: Why does this need to be here for connections?

File: src/adapter/src/rbac.rs [hunk 5]
```diff
@@ -418,6 +448,417 @@ fn ownership_err(
...
+            AclMode::CREATE,
+            role_id,
+        )]),
+        Plan::CreateSink(plan) => {
+            let mut privileges = BTreeSet::from([(
+                plan.name.qualifiers.clone().into(),
+                AclMode::CREATE,
+                role_id,
+            )]);
+            if let SourceSinkClusterConfig::Existing { id } = &plan.cluster_config {
```
Comment (mjibson) [inline src/adapter/src/rbac.rs:514]: I'm a bit scared of manually stuffing variants here and above. Does it make sense to have some named functions on `SourceSinkClusterConfig` that make the intention clearer?

File: src/adapter/src/rbac.rs [hunk 6]
```diff
@@ -418,6 +448,417 @@ fn ownership_err(
...
+fn generate_read_privileges_inner(
+    catalog: &impl SessionCatalog,
+    ids: impl Iterator<Item = GlobalId>,
+    role_id: RoleId,
+    seen: &mut BTreeSet<(GlobalId, RoleId)>,
+) -> BTreeSet<(ObjectId, AclMode, RoleId)> {
+    let mut privileges = BTreeSet::new();
+    for id in ids {
+        if !seen.contains(&(id, role_id)) {
+            seen.insert((id, role_id));
```
Comment (mjibson) [inline src/adapter/src/rbac.rs:793]:
```suggestion
        if seen.insert((id, role_id)) {
```

Unsure if this is clearer or not? Ignore if you think it's fine as is.

**Review by mjibson — APPROVED** [review]

File: src/sql/src/plan.rs (hunk 1 above)
Comment (jkosh44) [reply src/sql/src/plan.rs:108]: We need to know the schema of the object we're looking at so that we can check that the user has `SCHEMA(U)` privileges. So I created a new plan for `ShowCreate`, `ShowCreatePlan`, that included the `GlobalId` of the object we're looking at. After that `SendRowsPlan` was not used anywhere so I deleted it.

File: src/adapter/src/rbac.rs (hunk 3 above)
Comment (jkosh44) [reply src/adapter/src/rbac.rs:217]: Yes, this should probably go in another commit. I was just being lazy. I'll extract this into a separate PR.

File: src/adapter/src/rbac.rs (hunk 2 above)
Comment (jkosh44) [reply src/adapter/src/rbac.rs:221]: The `Existing` variant will create the source/sink in an existing cluster. So it's not creating a new cluster therefore it doesn't need the `CREATECLUSTER` attribute. It is creating a new object in a cluster though so it needs the `CLUSTER(U)` privilege on that specific cluster.

File: src/adapter/src/rbac.rs (hunk 4 above)
Comment (jkosh44) [reply src/adapter/src/rbac.rs:469]: Connections can use other connections and secrets. `generate_item_usage_privileges` works on secrets, connections, and types. Maybe that's too general? Do you think I should create three different functions so it's more explicit?

File: src/adapter/src/rbac.rs (hunk 5 above)
Comment (jkosh44) [reply src/adapter/src/rbac.rs]: I think that makes sense, I added a function and used that. Let me know what you think.

File: src/adapter/src/rbac.rs (hunk 6 above)
Comment (jkosh44) [reply src/adapter/src/rbac.rs]: I think your approach is a bit cleaner, I just updated it.

File: src/adapter/src/rbac.rs (hunk 4 above)
Comment (mjibson) [reply src/adapter/src/rbac.rs:448]: Nah, I think it's fine as is.

**Review by mjibson — APPROVED** [review]
Looks great!
